### ISSUE TITLE
[Feature] Add support to attach to running container terminal

### ISSUE THREAD

**Issue by patcfly** [mbecker20/komodo#75]
I would like to be able to attach to a running container's terminal in the Komodo web ui.

**Comment by mbecker20** [comment]
I'm not entirely sure how to accomplish this task.

I do have some idea to kind of fake it, achieving something close to an attach using a terminal client interface, and sending every line user executes atomically to `docker exec`.

I think this docker exec feature is much simpler, and should be able to do everything needed. But it's not docker attach and won't be just like docker attach over ssh. Mainly, the responses won't be stream back, but rather will resolve fully and send back whole response at once. The different wouldn't be noticeable for short running tasks but for anything long running, would give no feedback. 

Let me know if this docker exec feature would be useful to you, as a precursor to potential full docker attach with streaming responses

**Comment by phillibl** [comment]
I think having a button in komodo that runs something like 'docker exec -it mycontainer sh' for whatever the container is would be hugely beneficial.  

It's easy enough to run over ssh but having the feature in the komodo web ui would be great

**Comment by markhaines** [comment]
Would also love to see this. Dockge offers this ability; you can connect directly to the containers shell and it lets you switch from bash to sh if bash isn't supported.  It's super useful and one of the main things missing for me that stops me migrating from dockge solution.

**Comment by alphaleonis** [comment]
Same thing here. It is a really useful feature, so would love to see this implemented.

**Comment by SinisterSpatula** [comment]
+1

**Comment by henrysaantos** [comment]
+1

**Comment by flintstones-fred** [comment]
+1

**Comment by kingp0dd** [comment]
> I'm not entirely sure how to accomplish this task.
> 
> I do have some idea to kind of fake it, achieving something close to an attach using a terminal client interface, and sending every line user executes atomically to `docker exec`.
> 
> I think this docker exec feature is much simpler, and should be able to do everything needed. But it's not docker attach and won't be just like docker attach over ssh. Mainly, the responses won't be stream back, but rather will resolve fully and send back whole response at once. The different wouldn't be noticeable for short running tasks but for anything long running, would give no feedback.
> 
> Let me know if this docker exec feature would be useful to you, as a precursor to potential full docker attach with streaming responses

Docker exec one liners will be useful as i think majority of the use cases is this, for debugging or for executing quick backend commands. 

I'm not sure if that's what dockge also does. For portainer, i think it's a full attach?

**Comment by alphaleonis** [comment]
I'm no expert in this, but found the [following issue on github](https://github.com/portainer/portainer/issues/2640) which at least seems to contain some hints on how to accomplish this? (eg. running an interactive shell in a container using the API)

**Comment by stonkage** [comment]
+1

**Comment by nanosonde** [comment]
Hi!
Maybe you could use ttyd:
https://github.com/tsl0922/ttyd

See the usage examples in the ttyd wiki.

You could run something like:
ttyd <ttyd_options> docker exec -it container bash
It will then start a new web terminal with this command.

**Comment by nanosonde** [comment]
Also found this: https://lib.rs/crates/webterm

Also uses xterm.js, backend written in Rust...

**Comment by nanosonde** [comment]
Proxmox host and container consoles also use xterm.js:

https://github.com/proxmox/pve-xtermjs

**Comment by mac-tron** [comment]
+1

**Comment by d4rkw1n9** [comment]
...just to upvote for this crucial feature: +1

**Comment by Dinth** [comment]
+1

**Comment by RedVelocity** [comment]
Just following up on this topic. Is there a roadmap for the exec feature?

**Comment by Skyfay** [comment]
> Just following up on this topic. Is there a roadmap for the exec feature? 

As I see it, the problem is that the developer does not know how to implement this task at the moment.

**Comment by AlphaCraft9658** [comment]
+1

**Comment by lnlyssg** [comment]
Please stop commenting +1 - it clutters up the issue and stops any actual useful discussion from being easily seen. If you just want to upvote this then please add a thumbs-up reaction to the issue instead.

**Comment by lnlyssg** [comment]
It's coming! https://github.com/moghtech/komodo/pull/446

**Comment by Skyfay** [comment]
> It's coming! [#446](https://github.com/moghtech/komodo/pull/446)

This will initially only be the implementation for shell access. Exec still has to be executed manually. But Exec will most certainly be built on top of that.

**Comment by mbecker20** [comment]
container exec should be simple on top of the shell access in #446 . I'm going to go ahead and implement it before releasing 1.17.4

**Comment by nic96** [comment]
Wouldn't `docker attach` be just as simple as `docker exec -it` now that you have a terminal feature? I tested that it works attaching with the terminal feature e.g `docker attach --detach-keys=ctrl-d service-name`. 

The tty and std_open options need to be in the docker compose service to be able to attach and interact:
```
services:
  hello_world:
    image: hello-world
    tty: true
    stdin_open: true
```

**Comment by mbecker20** [comment]
@nic96 yes I am aware and attach should be fairly straightforward, leaving this issue open to implement this feature.

**Comment by Almie** [comment]
+1 to also having a docker attach feature, would be greatly beneficial for me.

**Comment by mbecker20** [comment]
@Almie it is available in the dev releases https://github.com/moghtech/komodo/releases
